internal/cmdset/cmdset.go

```diff
@@ -0,0 +1,125 @@
+		Args:    cobra.MaximumNArgs(1),
+		PreRunE: r.preRunE,
+		RunE:    r.runE,
+	}
+	runner.FixDocs(parent, c)
+	r.Command = c
+	c.Flags().StringVar(&r.KeyValue, "value", "",
+		"optional flag, the value of the setter to be set to")
+	c.Flags().StringVar(&r.SetBy, "set-by", "",
+		"annotate the field with who set it")
+	c.Flags().StringVar(&r.Description, "description", "",
+		"annotate the field with a description of its value")
+	c.Flags().BoolVarP(&r.RecurseSubPackages, "recurse-subpackages", "R", false,
+		"sets recursively in all the nested subpackages")
+
+	return r
+}
+
+func SetCommand(parent string) *cobra.Command {
+	return NewSetRunner(parent).Command
+}
+
+type SetRunner struct {
+	Command            *cobra.Command
+	Set                setters.FieldSetter
+	SetBy              string
+	Description        string
+	KeyValue           string
+	Name               string
+	Value              string
+	ListValues         []string
+	RecurseSubPackages bool
+}
+
+func (r *SetRunner) preRunE(c *cobra.Command, args []string) error {
+	valueFlagSet := c.Flag("value").Changed
+
+	if !valueFlagSet {
+		return errors.Errorf("value flag must be set")
+	}
+
+	// r.KeyValue holds SETTER_NAME=SETTER_VALUE, split them into name and value
+	r.Name = strings.Split(r.KeyValue, "=")[0]
+	setterValue := strings.Split(r.KeyValue, "=")[1]
+
+	// if values is a list e.g. [a, b, c], this is array input, so populate r.Listvalues
+	// instead
+	if strings.HasPrefix(setterValue, "[") && strings.HasSuffix(setterValue, "]") {
+		commaSepVals := strings.TrimSuffix(strings.TrimPrefix(setterValue, "["), "]")
+		r.ListValues = strings.Split(commaSepVals, ",")
+		for i := range r.ListValues {
+			r.ListValues[i] = strings.TrimSpace(r.ListValues[i])
+		}
+	} else {
+		// else it is a scalar value
+		r.Value = setterValue
+	}
+	return nil
+}
+
+func (r *SetRunner) runE(c *cobra.Command, args []string) error {
+	e := runner.ExecuteCmdOnPkgs{
+		NeedOpenAPI:        true,
+		Writer:             c.OutOrStdout(),
+		RootPkgPath:        args[0],
+		RecurseSubPackages: r.RecurseSubPackages,
+		CmdRunner:          r,
+	}
+	err := e.Execute()
+	if err != nil {
+		return runner.HandleError(c, err)
+	}
+	return nil
+}
+
+func (r *SetRunner) ExecuteCmd(w io.Writer, pkgPath string) error {
+	r.Set = setters.FieldSetter{
+		Name:               r.Name,
+		Value:              r.Value,
+		ListValues:         r.ListValues,
+		Description:        r.Description,
+		SetBy:              r.SetBy,
+		Count:              0,
+		OpenAPIPath:        filepath.Join(pkgPath, ext.KRMFileName()),
+		OpenAPIFileName:    ext.KRMFileName(),
+		ResourcesPath:      pkgPath,
+		RecurseSubPackages: r.RecurseSubPackages,
+		IsSet:              true,
+	}
+	count, err := r.Set.Set()
+	if err != nil {
+		// return err if RecurseSubPackages is false
+		if !r.Set.RecurseSubPackages {
+			return err
+		}
+		// print error message and continue if RecurseSubPackages is true
+		fmt.Fprintf(w, "%s\n", err.Error())
```

Comment:
Should we try to separate executing the command from printing? It doesn't matter that much in this example, but this is an area where I think we want to do better than the existing commands.

Reply:
ExecuteCmd method at cli layer will only print the out message here. The libraries won't print any messages. I think this is simple and good enough for commands like this. What do you mean by separating further ? Wrap it in a different method?

Reply:
This solution is certainly much better than printing within the library. It still requires us to print for each subpackage individually. It is a step in the right direction.